### DISCUSSION TITLE
useSimulateContract - Slow and unexpected return

### DISCUSSION THREAD

**Discussion by LeoClairet05** [wevm/wagmi#3474] [Q&A]
### Describe the bug

I am trying to catch error in simulation, I am using "error" return for that.

In the below code: "depositAmountValue" is a state modifiable from an input. For my test I intentionally try to deposit more than my balance. 
Expected return of error would be something like: "ERC20 amount exceed balance"

I get null first, then after useSimulateContract finish loading (isLoading == false) I get the expected error and then automatically without any changes the error switch back to null.

+ useSimulateContract is very slow compared to wagmi v1. (Take around 10-20 seconds before finish loading)
+ isError is always false

```
const { data: DepositConfig, error: _depositError, isLoading } = useSimulateContract(
        {
            ...depositContract,
            functionName: "deposit",
            args: [ depositAmountValue ],
            query: {
                enabled: depositAmountValue != "0"
            }
        }
  )

useEffect(()=>{
        console.log(_depositError, DepositToVaultConfig, isError)
        if(_depositError && !DepositToVaultConfig){
            
            setDepositError(_depositError?.shortMessage);
        }else{
            setDepositError(null)
        }

    }, [_depositError, DepositToVaultConfig, isLoading])
```

### Wagmi Version

2.2.1

### Viem Version

2.1.1

### Check existing issues

- [X] I checked there isn't [already an issue](https://github.com/wevm/wagmi/issues) for the bug I encountered.

**Comment by 0xMySt1c** [comment]
BUMP.
having similar issues
the errors are returning very slowly

**Comment by songkeys** [comment]
~~Because when you are typing, it triggers lots of requests. It seems these requests are blocking the following ones. And you have to wait for the last one to finish. I think a debounce should be implemented in the library. Also the following requests should cancel the previous ones.~~

~~Before this is resolved, you can also resolve this in your app end (with a debounced state arg).~~

edit: okay sorry I put up the solution without even trying it. after trying with debounced solution, it still sending several requests even when the changes are only one. It looks like the input body is the same but the `id` is increasing for serveral times.

<img width="233" alt="SCR-20240130-bryo" src="https://github.com/wevm/wagmi/assets/22665058/476fa954-6ecb-4d35-adeb-50123af3813a">

edit again: okay i think i guess i've figured out this bug - it's trying to retry because the response is not right. adding `query: { retry: false }` could solve but there has to be a better way.

Btw, the `ChainNotConfiguredError` is always thrown on first tryings because of this line i think:

https://github.com/wevm/wagmi/blob/f9ed00ec39c71648a44dbcc3a7ce21824a0e0669/packages/react/src/hooks/useSimulateContract.ts#L97

or this? i'm not sure:

https://github.com/wevm/wagmi/blob/f9ed00ec39c71648a44dbcc3a7ce21824a0e0669/packages/react/src/hooks/useSimulateContract.ts#L101

Then if the input triggers `ContractFunctionExecutionError`, it'll retry for several times because it considers the error as a network error though it is not actually.

**Reply by 0xMySt1c** [reply]
hey so setting retry to false did seem to fix this issue. thanks for fixing this. have you found a more elegant way to fix this? do you think this maybe an issue with how simulateContract is implemented?

**Reply by 0xTARC** [reply]
There is no implementation issue in wagmi, unless you count the retries being set to 3 by default.

You can easily disable this behavior globally by modifying your `QueryClient` like so:


```
      new QueryClient({
        defaultOptions: {
          queries: {
            retry: false,
            ...
          },
        },
     })
```
